README.md
[![NPM version](https://badge.fury.io/js/cdk-fargate-run-task.svg)](https://badge.fury.io/js/cdk-fargate-run-task)
[![PyPI version](https://badge.fury.io/py/cdk-fargate-run-task.svg)](https://badge.fury.io/py/cdk-fargate-run-task)
![Release](https://github.com/pahud/cdk-fargate-run-task/workflows/Release/badge.svg?branch=main)

# cdk-fargate-run-task

Define and run container tasks on AWS Fargate at once or by schedule.

# sample

```ts
const app = new cdk.App();

const env = {
  account: process.env.CDK_DEFAULT_ACCOUNT,
  region: process.env.CDK_DEFAULT_REGION,
};

const stack = new cdk.Stack(app, 'run-task-demo-stack', { env });

// define your task
const task = new ecs.FargateTaskDefinition(stack, 'Task', { cpu: 256, memoryLimitMiB: 512 });

// add contianer into the task
task.addContainer('Ping', {
  image: ecs.ContainerImage.fromRegistry('busybox'),
  command: [
    'sh', '-c',
    'ping -c 3 google.com',
  ],
  logging: new ecs.AwsLogDriver({
    streamPrefix: 'Ping',
    logGroup: new LogGroup(stack, 'LogGroup', {
      logGroupName: `${stack.stackName}LogGroup`,
      retention: RetentionDays.ONE_DAY,
    }),
  }),
});

// deploy and run this task once
const runTaskAtOnce = new RunTask(stack, 'RunDemoTaskOnce', { task });

// or run it with schedule(every hour 0min)
new RunTask(stack, 'RunDemoTaskEveryHour', {
  task,
  cluster: runTaskAtOnce.cluster,
  runOnce: false,
  schedule: Schedule.cron({ minute: '0' }),
});

```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.1.16](https://github.com/pahud/cdk-fargate-run-task/compare/v0.1.15...v0.1.16) (2021-04-21)

### [0.1.15](https://github.com/pahud/cdk-fargate-run-task/compare/v0.1.14...v0.1.15) (2021-04-20)

### [0.1.14](https://github.com/pahud/cdk-fargate-run-task/compare/v0.1.13...v0.1.14) (2021-04-19)

### [0.1.13](https://github.com/pahud/cdk-fargate-run-task/compare/v0.1.12...v0.1.13) (2021-04-18)

### [0.1.12](https://github.com/pahud/cdk-fargate-run-task/compare/v0.1.11...v0.1.12) (2021-04-17)

### [0.1.11](https://github.com/pahud/cdk-fargate-run-task/compare/v0.1.10...v0.1.11) (2021-04-16)

### [0.1.10](https://github.com/pahud/cdk-fargate-run-task/compare/v0.1.9...v0.1.10) (2021-04-15)

### [0.1.9](https://github.com/pahud/cdk-fargate-run-task/compare/v0.1.8...v0.1.9) (2021-04-14)

### [0.1.8](https://github.com/pahud/cdk-fargate-run-task/compare/v0.1.7...v0.1.8) (2021-04-13)

### [0.1.7](https://github.com/pahud/cdk-fargate-run-task/compare/v0.1.6...v0.1.7) (2021-04-12)

### [0.1.6](https://github.com/pahud/cdk-fargate-run-task/compare/v0.1.5...v0.1.6) (2021-04-11)

### [0.1.5](https://github.com/pahud/cdk-fargate-run-task/compare/v0.1.4...v0.1.5) (2021-04-10)

### [0.1.4](https://github.com/pahud/cdk-fargate-run-task/compare/v0.1.3...v0.1.4) (2021-04-09)

### [0.1.3](https://github.com/pahud/cdk-fargate-run-task/compare/v0.1.2...v0.1.3) (2021-04-08)

### [0.1.2](https://github.com/pahud/cdk-fargate-run-task/compare/v0.1.1...v0.1.2) (2021-04-07)

### [0.1.1](https://github.com/pahud/cdk-fargate-run-task/compare/v0.1.0...v0.1.1) (2021-04-02)

## [0.1.0](https://github.com/pahud/cdk-fargate-run-task/compare/v0.0.82...v0.1.0) (2021-04-02)


### Features

* support fargate spot ([#83](https://github.com/pahud/cdk-fargate-run-task/issues/83)) ([2b09885](https://github.com/pahud/cdk-fargate-run-task/commit/2b098855a725875de3b0f69ccf05c0fe496f585d))

### [0.0.82](https://github.com/pahud/cdk-fargate-run-task/compare/v0.0.81...v0.0.82) (2021-04-02)

### [0.0.81](https://github.com/pahud/cdk-fargate-run-task/compare/v0.0.80...v0.0.81) (2021-04-01)

### [0.0.80](https://github.com/pahud/cdk-fargate-run-task/compare/v0.0.79...v0.0.80) (2021-03-31)

### [0.0.79](https://github.com/pahud/cdk-fargate-run-task/compare/v0.0.78...v0.0.79) (2021-03-30)

### [0.0.78](https://github.com/pahud/cdk-fargate-run-task/compare/v0.0.77...v0.0.78) (2021-03-29)

### [0.0.77](https://github.com/pahud/cdk-fargate-run-task/compare/v0.0.76...v0.0.77) (2021-03-27)

### [0.0.76](https://github.com/pahud/cdk-fargate-run-task/compare/v0.0.75...v0.0.76) (2021-03-26)

### [0.0.75](https://github.com/pahud/cdk-fargate-run-task/compare/v0.0.74...v0.0.75) (2021-03-25)

### [0.0.74](https://github.com/pahud/cdk-fargate-run-task/compare/v0.0.73...v0.0.74) (2021-03-24)

### [0.0.73](https://github.com/pahud/cdk-fargate-run-task/compare/v0.0.72...v0.0.73) (2021-03-22)

### [0.0.72](https://github.com/pahud/cdk-fargate-run-task/compare/v0.0.71...v0.0.72) (2021-03-21)

### [0.0.71](https://github.com/pahud/cdk-fargate-run-task/compare/v0.0.70...v0.0.71) (2021-03-20)

### [0.0.70](https://github.com/pahud/cdk-fargate-run-task/compare/v0.0.69...v0.0.70) (2021-03-19)

### [0.0.69](https://github.com/pahud/cdk-fargate-run-task/compare/v0.0.68...v0.0.69) (2021-03-18)

### [0.0.68](https://github.com/pahud/cdk-fargate-run-task/compare/v0.0.67...v0.0.68) (2021-03-06)

### [0.0.67](https://github.com/pahud/cdk-fargate-run-task/compare/v0.0.66...v0.0.67) (2021-03-05)

### [0.0.66](https://github.com/pahud/cdk-fargate-run-task/compare/v0.0.65...v0.0.66) (2021-03-04)

### [0.0.65](https://github.com/pahud/cdk-fargate-run-task/compare/v0.0.64...v0.0.65) (2021-03-03)

### [0.0.64](https://github.com/pahud/cdk-fargate-run-task/compare/v0.0.63...v0.0.64) (2021-03-03)

### [0.0.63](https://github.com/pahud/cdk-fargate-run-task/compare/v0.0.62...v0.0.63) (2021-03-03)


### Bug Fixes

* update platform version number ([#62](https://github.com/pahud/cdk-fargate-run-task/issues/62)) ([bf27f29](https://github.com/pahud/cdk-fargate-run-task/commit/bf27f2907d3015fad3af1277dca482f2c72d4e8f))

### [0.0.62](https://github.com/pahud/cdk-fargate-run-task/compare/v0.0.61...v0.0.62) (2021-03-03)

### [0.0.61](https://github.com/pahud/cdk-fargate-run-task/compare/v0.0.60...v0.0.61) (2021-03-03)

### [0.0.60](https://github.com/pahud/cdk-fargate-run-task/compare/v0.0.59...v0.0.60) (2021-03-03)

### [0.0.59](https://github.com/pahud/cdk-fargate-run-task/compare/v0.0.58...v0.0.59) (2021-03-02)

### [0.0.58](https://github.com/pahud/cdk-fargate-run-task/compare/v0.0.57...v0.0.58) (2021-03-01)

### [0.0.57](https://github.com/pahud/cdk-fargate-run-task/compare/v0.0.56...v0.0.57) (2021-02-28)

### [0.0.56](https://github.com/pahud/cdk-fargate-run-task/compare/v0.0.55...v0.0.56) (2021-02-27)

### [0.0.55](https://github.com/pahud/cdk-fargate-run-task/compare/v0.0.54...v0.0.55) (2021-02-26)

### [0.0.54](https://github.com/pahud/cdk-fargate-run-task/compare/v0.0.53...v0.0.54) (2021-02-25)

### [0.0.53](https://github.com/pahud/cdk-fargate-run-task/compare/v0.0.52...v0.0.53) (2021-02-24)

### [0.0.52](https://github.com/pahud/cdk-fargate-run-task/compare/v0.0.51...v0.0.52) (2021-02-23)

### [0.0.51](https://github.com/pahud/cdk-fargate-run-task/compare/v0.0.50...v0.0.51) (2021-02-21)

### [0.0.50](https://github.com/pahud/cdk-fargate-run-task/compare/v0.0.49...v0.0.50) (2021-02-20)

### [0.0.49](https://github.com/pahud/cdk-fargate-run-task/compare/v0.0.48...v0.0.49) (2021-02-19)

### [0.0.48](https://github.com/pahud/cdk-fargate-run-task/compare/v0.0.47...v0.0.48) (2021-02-18)

### [0.0.47](https://github.com/pahud/cdk-fargate-run-task/compare/v0.0.46...v0.0.47) (2021-02-17)

### [0.0.46](https://github.com/pahud/cdk-fargate-run-task/compare/v0.0.45...v0.0.46) (2021-02-16)


.projenrc.js
const { AwsCdkConstructLibrary } = require('projen');

const AUTOMATION_TOKEN = 'AUTOMATION_GITHUB_TOKEN';

const project = new AwsCdkConstructLibrary({
  authorAddress: 'pahudnet@gmail.com',
  authorName: 'Pahud',
  cdkVersion: '1.73.0',
  name: 'cdk-fargate-run-task',
  releaseBranches: ['main'],
  defaultReleaseBranch: ['main'],
  description: 'Define and run container tasks on AWS Fargate immediately or with schedule',
  repository: 'https://github.com/pahud/cdk-fargate-run-task.git',
  cdkDependencies: [
    '@aws-cdk/core',
    '@aws-cdk/aws-ec2',
    '@aws-cdk/aws-ecs',
    '@aws-cdk/aws-events-targets',
    '@aws-cdk/aws-events',
    '@aws-cdk/aws-logs',
    '@aws-cdk/custom-resources',
  ],
  catalog: {
    announce: false,
    twitter: 'pahudnet',
  },
  python: {
    distName: 'cdk-fargate-run-task',
    module: 'cdk_fargate_run_task',
  },
  dependabot: false,
  // upgrade every Sunday 6AM
  projenUpgradeSchedule: ['0 6 * * 0'],
  keywords: [
    'cdk',
    'fargate',
    'task',
    'ecs',
    'aws',
  ],

});


// create a custom projen and yarn upgrade workflow
const workflow = project.github.addWorkflow('ProjenYarnUpgrade');

workflow.on({
  schedule: [{
    cron: '11 0 * * *',
  }], // 0:11am every day
  workflow_dispatch: {}, // allow manual triggering
});

workflow.addJobs({
  upgrade: {
    'runs-on': 'ubuntu-latest',
    'steps': [
      { uses: 'actions/checkout@v2' },
      {
        uses: 'actions/setup-node@v1',
        with: {
          'node-version': '10.17.0',
        },
      },
      { run: 'yarn upgrade' },
      { run: 'yarn projen:upgrade' },
      // submit a PR
      {
        name: 'Create Pull Request',
        uses: 'peter-evans/create-pull-request@v3',
        with: {
          'token': '${{ secrets.' + AUTOMATION_TOKEN + ' }}',
          'commit-message': 'chore: upgrade projen',
          'branch': 'auto/projen-upgrade',
          'title': 'chore: upgrade projen and yarn',
          'body': 'This PR upgrades projen and yarn upgrade to the latest version',
          'labels': 'auto-merge',
        },
      },
    ],
  },
});

const common_exclude = ['cdk.out', 'cdk.context.json', 'images', 'yarn-error.log'];
project.npmignore.exclude(...common_exclude);
project.gitignore.exclude(...common_exclude);

project.synth();


src/run-task.ts
import * as ec2 from '@aws-cdk/aws-ec2';
import * as ecs from '@aws-cdk/aws-ecs';
import { Rule, Schedule } from '@aws-cdk/aws-events';
import { EcsTask } from '@aws-cdk/aws-events-targets';
import { RetentionDays } from '@aws-cdk/aws-logs';
import { Construct } from '@aws-cdk/core';
import * as cr from '@aws-cdk/custom-resources';


export interface RunTaskProps {
  /**
   * The VPC for the Amazon ECS task
   * @default - create a new VPC or use existing one
   */
  readonly vpc?: ec2.IVpc;
  /**
   * The Amazon ECS Cluster
   * @default - create a new cluster
   */
  readonly cluster?: ecs.ICluster;
  /**
   * The Amazon ECS Task definition for AWS Fargate
   */
  readonly task: ecs.FargateTaskDefinition;
  /**
   * run it at once(immediately after deployment)
   * @default true
   */
  readonly runAtOnce?: boolean;
  /**
   * run the task with defined schedule
   * @default - no shedule
   */
  readonly schedule?: Schedule;
  /**
   * Log retention days
   * @default - one week
   */
  readonly logRetention?: RetentionDays;
}

export class RunTask extends Construct {
  readonly vpc: ec2.IVpc;
  readonly cluster: ecs.ICluster;
  constructor(scope: Construct, id: string, props: RunTaskProps) {
    super(scope, id);

    // const stack = Stack.of(this);
    const vpc = props.vpc ?? props.cluster ? props.cluster!.vpc : getOrCreateVpc(this);
    const cluster = props.cluster ?? new ecs.Cluster(this, 'Cluster', { vpc });
    const task = props.task;
    this.vpc = vpc;
    this.cluster = cluster;

    if (props.schedule) {
      new Rule(this, 'ScheduleRule', {
        schedule: props.schedule,
        targets: [
          new EcsTask({
            cluster,
            taskDefinition: task,
          }),
        ],
      });
    }

    if (props.runAtOnce !== false) {
      const onEvent = {
        service: 'ECS',
        action: 'runTask',
        parameters: {
          cluster: cluster.clusterName,
          taskDefinition: task.taskDefinitionArn,
          launchType: 'FARGATE',
          networkConfiguration: {
            awsvpcConfiguration: {
              assignPublicIp: 'DISABLED',
              subnets: vpc.selectSubnets({
                subnetType: ec2.SubnetType.PRIVATE,
              }).subnetIds,
            },
          },
        },
        physicalResourceId: cr.PhysicalResourceId.of(task.taskDefinitionArn),
      };
      const runTaskResource = new cr.AwsCustomResource(this, 'EcsRunTask', {
        onCreate: onEvent,
        onUpdate: onEvent,
        policy: cr.AwsCustomResourcePolicy.fromSdkCalls({ resources: [task.taskDefinitionArn] }),
        logRetention: props.logRetention ?? RetentionDays.ONE_WEEK,
      });

      // allow lambda from custom resource to iam:PassRole on the ecs task role and execution role
      task.taskRole.grantPassRole(runTaskResource.grantPrincipal);
      if (task.executionRole) task.executionRole.grantPassRole(runTaskResource.grantPrincipal);
      // new CfnOutput(stack, `TaskDefinitionArn-${id}`, { value: task.taskDefinitionArn });
    }

  }
}

function getOrCreateVpc(scope: Construct): ec2.IVpc {
  // use an existing vpc or create a new one
  return scope.node.tryGetContext('use_default_vpc') === '1' ?
    ec2.Vpc.fromLookup(scope, 'Vpc', { isDefault: true }) :
    scope.node.tryGetContext('use_vpc_id') ?
      ec2.Vpc.fromLookup(scope, 'Vpc', { vpcId: scope.node.tryGetContext('use_vpc_id') }) :
      new ec2.Vpc(scope, 'Vpc', { maxAzs: 3, natGateways: 1 });
}


src/integ.default.ts
import * as ecs from '@aws-cdk/aws-ecs';
import { Schedule } from '@aws-cdk/aws-events';
import { LogGroup, RetentionDays } from '@aws-cdk/aws-logs';
import * as cdk from '@aws-cdk/core';
import { RunTask } from './run-task';

export class IntegTesting {
  readonly stack: cdk.Stack[];
  constructor() {
    const app = new cdk.App();

    const env = {
      account: process.env.CDK_DEFAULT_ACCOUNT,
      region: process.env.CDK_DEFAULT_REGION,
    };

    const stack = new cdk.Stack(app, 'run-task-demo-stack', { env });

    const task = new ecs.FargateTaskDefinition(stack, 'Task', { cpu: 256, memoryLimitMiB: 512 });

    task.addContainer('Ping', {
      image: ecs.ContainerImage.fromRegistry('busybox'),
      command: [
        'sh', '-c',
        'ping -c 3 google.com',
      ],
      logging: new ecs.AwsLogDriver({
        streamPrefix: 'Ping',
        logGroup: new LogGroup(stack, 'LogGroup', {
          logGroupName: `${stack.stackName}LogGroup`,
          retention: RetentionDays.ONE_DAY,
        }),
      }),
    });

    // deploy and run this task once
    const runTaskAtOnce = new RunTask(stack, 'RunDemoTaskOnce', { task });

    // or run it with schedule(every hour 0min)
    new RunTask(stack, 'RunDemoTaskEveryHour', {
      task,
      cluster: runTaskAtOnce.cluster,
      runAtOnce: false,
      schedule: Schedule.cron({ minute: '0' }),
    });

    app.synth();
    this.stack = [stack];
  }
}

new IntegTesting;
